include/dlaf/communication/communicator_pipeline.h

```diff
@@ -0,0 +1,141 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2023, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+/// @file
+
+#pragma once
+
+#include <dlaf/common/index2d.h>
+#include <dlaf/common/pipeline.h>
+#include <dlaf/common/round_robin.h>
+#include <dlaf/communication/communicator.h>
+#include <dlaf/communication/communicator_type.h>
+#include <dlaf/communication/index.h>
+
+namespace dlaf::comm {
+namespace internal {
+constexpr const dlaf::common::Ordering FULL_COMMUNICATOR_ORDER{dlaf::common::Ordering::RowMajor};
+}
+
+using CommunicatorPipelineReadOnlyWrapper = typename common::Pipeline<Communicator>::ReadOnlyWrapper;
+using CommunicatorPipelineReadWriteWrapper = typename common::Pipeline<Communicator>::ReadWriteWrapper;
+using CommunicatorPipelineReadOnlySender = typename common::Pipeline<Communicator>::ReadOnlySender;
+using CommunicatorPipelineReadWriteSender = typename common::Pipeline<Communicator>::ReadWriteSender;
+
+/// A CommunicatorPipeline provides pipelined access to a communicator, as well as metadata about the
+/// communicator.
+///
+/// A CommunicatorPipeline uses a Pipeline to provide shared or exclusive access through senders to the
+/// underlying communicator.  @tparam CT signals whether the contained communicator is a row, column, or
+/// full grid communicator.
+template <CommunicatorType CT>
+class CommunicatorPipeline {
+  using PipelineType = dlaf::common::Pipeline<Communicator>;
+
+public:
+  using ReadOnlyWrapper = typename PipelineType::ReadOnlyWrapper;
+  using ReadWriteWrapper = typename PipelineType::ReadWriteWrapper;
+  using ReadOnlySender = typename PipelineType::ReadOnlySender;
+  using ReadWriteSender = typename PipelineType::ReadWriteSender;
+
+  static constexpr CommunicatorType communicator_type = CT;
+
+  /// Create a CommunicatorPipeline by moving in the resource (it takes the
+  /// ownership).
+  explicit CommunicatorPipeline(Communicator comm, Index2D rank = {0, 0}, Size2D size = {0, 0})
```

Comment:
I've updated both to default to `{-1, -1}`. However, I noticed that `Index2D` defaults to `{-1, -1}` while `Size2D` doesn't have a default constructor. Do we want to default the size (here or in `index2d.h`) to `{0, 0}` instead?

Reply:
IMO `{-1, -1}` is more meaningful than `{0, 0}`, since it immediately tells you that the coordinate is invalid and, AFAIU, when you pass a Communicator not obtained from a Grid, both `rank` and `size` should not be used at all, and being invalid feel safer to me.

Reply:
All right, well they're both `{-1, -1}` now so marking this as resolved in that case.